### PR TITLE
Add Venue booking feature

### PR DESCRIPTION
**Pull request by nseah21**
# Feature has been completed
## Things to take note:
* Tests update have been delayed to v1.4 under #179
* UI update has mostly been done, just need some prettifying under #180

### CODE REVIEW

**Review by nseah21 — COMMENTED** [review]
Good work. Please see my comments below.

File: src/main/java/seedu/rc4hdb/model/venues/exceptions/VenueNotFoundException.java [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package seedu.rc4hdb.model.venues.exceptions;
+
+/**
+ * Signals that the operation is unable to find the specified venue.
+ */
+public class VenueNotFoundException extends RuntimeException {
+}
```
[the remark this answers is not in the document]
Comment (jq1836) [reply src/main/java/seedu/rc4hdb/model/venues/exceptions/VenueNotFoundException.java:7]: Let's bring this to v1.4 hahaha

File: src/main/java/seedu/rc4hdb/model/venues/UniqueVenueList.java [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+     */
+    private boolean venuesAreUnique(List<Venue> venues) {
+        for (int i = 0; i < venues.size() - 1; i++) {
+            for (int j = i + 1; j < venues.size(); j++) {
+                if (venues.get(i).isSameVenue(venues.get(j))) {
+                    return false;
+                }
+            }
+        }
+        return true;
```
Comment (nseah21) [inline src/main/java/seedu/rc4hdb/model/venues/UniqueVenueList.java:173]: Perhaps deep nesting can be avoided here through the use of streams?

File: src/main/java/seedu/rc4hdb/logic/commands/venuecommands/UnbookCommand.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+            return new CommandResult(String.format(MESSAGE_SUCCESS, venueName,
+                    bookingDescriptor.getDayOfWeek().orElse(null),
+                    bookingDescriptor.getHourPeriod().orElse(null)));
+        } catch (VenueNotFoundException e) {
+            throw new CommandException(String.format(MESSAGE_VENUE_NOT_FOUND, venueName));
+        } catch (BookingNotFoundException e) {
+            throw new CommandException(String.format(MESSAGE_BOOKING_NOT_FOUND, venueName,
+                    bookingDescriptor.getDayOfWeek().orElse(null),
+                    bookingDescriptor.getHourPeriod().orElse(null)), e);
+        }
```
[the remark this answers is not in the document]
Comment (jq1836) [reply src/main/java/seedu/rc4hdb/logic/commands/venuecommands/UnbookCommand.java:63]: I don't think so, the `VenueNotFoundException` and `BookingNotFoundException`s are thrown by `removeBooking` and `setObservableBookings` methods from `Model`. Though, I think that the `orElse(null)` portion is lazy implementation on my part. I was supposed to use `Optional::get` method to retrieve the information and handle the exception when `bookingDescriptor` does not contain the required fields.

**Review by nealetham — APPROVED** [review]
Approved on behalf of Nicholas.

File: src/main/java/seedu/rc4hdb/logic/commands/venuecommands/BookCommand.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+            throw new CommandException(MESSAGE_CLASHING_BOOKING, e);
+        }
+    }
+
+    /**
+     * Creates and returns a {@code Booking} with the details of {@code bookingDescriptor}.
+     */
+    private Booking createNewBooking(Resident resident) throws NoSuchElementException {
+        assert bookingDescriptor != null && resident != null;
+        HourPeriod hourPeriod = bookingDescriptor.getHourPeriod().get();
```
Comment (nealetham) [inline src/main/java/seedu/rc4hdb/logic/commands/venuecommands/BookCommand.java:88]: From what we've learnt in the textbook, does this violate `Law of Demeter`?

Comment (nealetham) [reply src/main/java/seedu/rc4hdb/logic/commands/venuecommands/BookCommand.java:88]: This repeats for line 89 and 90. Though we currently have no quick way to fix it.

**Review by nealetham — APPROVED** [review]
LGTM, just as you said, we can update the code quality after v1.3

**Pull request merged**